Improve the metric of shadow cache for relatively small working set

beinan opened the pull request:
### What changes are proposed in this pull request?

Update the shadow cache metrics in 'get' method

Expose shadow cache metrics by calling the incrementCounts method in CacheContext

### Why are the changes needed?
Without this change, the shadow cache metrics would be only updated in 'get' method, which might cause the size of working set inaccurate (far below the real value) especially when the working set is lower than the cache capacity.

Our customer can also get table level metrics by overriding the incrementCounts method

beinan commented:
alluxio-bot, check this please

beinan commented:
Tested with presto (time window is 4m).  We can see the shadow cache bytes (working set) got increased (second peak) when everything had been in cache.   And we can also see it dipped when no one read the data.

Shadow cache bytes (working set)
![image](https://user-images.githubusercontent.com/7077075/125735403-82b8a880-c933-4ff8-b302-726ab9298c52.png)
Bytes read from external
![image](https://user-images.githubusercontent.com/7077075/125735615-5878c9f1-a878-4d53-a265-ad75cee76538.png)

apc999 reviewed (changes requested):
the code change looks good to me.
Can you add unit test to cover these cases?

beinan commented:
> the code change looks good to me.
> Can you add unit test to cover these cases?

Hey @apc999 , I added a unit test.  
I also added another commit to expose shadow cache increment to cacheContext.  Could you pls take a look?  Thanks!

apc999 commented on core/client/fs/src/main/java/alluxio/client/file/cache/PageId.java:
```diff
@@ -23,14 +26,25 @@
 public class PageId {
   private final String mFileId;
   private final long mPageIndex;
+  private final Optional<CacheContext> mCacheContext;
```
why do we include CacheContext in PageId?

beinan replied:
Good question, I also struggled with this for a while.  I'm just thinking each page could contains a cacheContext, but I didn't see a data structure for the page.  I include the cacheContext in PageId, so that the cacheContext can be accessed much easier in all places where the page is operated(put/get/evict), especially when we implement TTL or some other eviction policy later.
 
If we don't include the CacheContext in PageId, we might need to modify the signature of the get and put methods to pass the cacheContext into CacheManager.  What do you think?

apc999 replied:
I am good to put CacheCotext into Put/Get signature ---- I think that's the correct

beinan replied:
Done.

Hi @dbw9580, I changed the signature of the get/put method in CacheManager,  it would be easier to get a cacheContext, but I guess it might cause lots of conflict with your local change.  Sorry for that :(

apc999 commented on core/client/fs/src/main/java/alluxio/client/file/cache/NoExceptionCacheManager.java:
```diff
@@ -80,6 +79,20 @@ public int get(PageId pageId, int pageOffset, int bytesToRead, byte[] buffer,
+      return mCacheManager
+          .get(pageId, pageOffset, bytesToRead, buffer, offsetInBuffer, cacheContext);
+    } catch (Exception e) {
+      LOG.error("Failed to get page {}, offset {} cacheContext", pageId, pageOffset, cacheContext,
```
```suggestion
      LOG.error("Failed to get page {}, offset {} cacheContext {}", pageId, pageOffset, cacheContext,
```

beinan replied:
done

apc999 commented on core/client/fs/src/main/java/alluxio/client/file/cache/CacheManagerWithShadowCache.java:
```diff
@@ -266,6 +271,7 @@ public void funnel(PageId from, PrimitiveSink into) {
   }
 
   private static final class Metrics {
+
```
nit: remove

beinan replied:
done

apc999 reviewed (changes requested):
the code becomes cleaner and easier to maintain after this refactor.
I left some minor comments but it looks good overall.

Thanks!

beinan commented on core/client/fs/src/main/java/alluxio/client/file/cache/CacheManagerWithShadowCache.java:
```diff
@@ -37,6 +36,7 @@
  * A wrapper class of CacheManager with shadow cache.
  */
 public class CacheManagerWithShadowCache implements CacheManager {
+
```
done

beinan commented on core/common/src/main/java/alluxio/client/file/CacheContext.java:
```diff
@@ -24,6 +24,9 @@
  */
 public class CacheContext {
+
+  public static final CacheContext DEFAULT_CACHE_CONTEXT = new CacheContext();
```
I removed the singleton default "DEFAULT_CACHE_CONTEXT" and use the defaults() method instead, because the cacheContext is mutable, so if somewhere modify the DEFAULT_CACHE_CONTEXT, it would cause unexpected behaviors in other places.   What do you think?

apc999 replied:
yes, it is the case I am concerning too.

apc999 commented:
```
Error: 0.917 [ERROR] /usr/src/alluxio/core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheManagerTest.java:[353,28] CacheContext() has private access in alluxio.client.file.CacheContext
Error: 0.917 [ERROR] /usr/src/alluxio/core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheManagerTest.java:[387,30] CacheContext() has private access in alluxio.client.file.CacheContext
Error: 0.918 [ERROR] /usr/src/alluxio/core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheManagerTest.java:[415,32] CacheContext() has private access in alluxio.client.file.CacheContext
Error: 0.918 [ERROR] /usr/src/alluxio/core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheManagerTest.java:[445,31] CacheContext() has private access in alluxio.client.file.CacheContext
Error: 0.919 [ERROR] /usr/src/alluxio/core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheManagerTest.java:[449,31] CacheContext() has private access in alluxio.client.file.CacheContext
```

apc999 commented on core/client/fs/src/test/java/alluxio/client/file/cache/LocalCacheManagerTest.java:
```diff
@@ -435,9 +442,13 @@ public void putWithInsufficientParentQuota() throws Exception {
+      assertTrue(mCacheManager.put(PAGE_ID1, PAGE1, context1));
       assertEquals(PAGE1.length, mCacheManager.get(PAGE_ID1, PAGE1.length, mBuf, 0));
-      assertTrue(mCacheManager.put(PAGE_ID2, PAGE2, partitionCacheScope2, quota));
+      CacheContext context2 = new CacheContext().setCacheScope(partitionCacheScope2)
```
```suggestion
      CacheContext context2 = CacheContext().defaults().setCacheScope(partitionCacheScope2)
```

apc999 reviewed (approved):
LGTM modulo the build

apc999 commented:
Error: 9.287 [ERROR] alluxio.client.file.cache.LocalCacheManagerTest.getOffset  Time elapsed: 0.123 s  <<< FAILURE!
java.lang.AssertionError: expected:<java.nio.HeapByteBuffer[pos=1 lim=1024 cap=1024]> but was:<java.nio.HeapByteBuffer[pos=0 lim=1023 cap=1024]>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at alluxio.client.file.cache.LocalCacheManagerTest.getOffset(LocalCacheManagerTest.java:490)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:364)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:237)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:158)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)

Error: 9.335 [ERROR] alluxio.client.file.cache.LocalCacheManagerTest.getOffset  Time elapsed: 0.234 s  <<< FAILURE!
java.lang.AssertionError: expected:<java.nio.HeapByteBuffer[pos=1 lim=1024 cap=1024]> but was:<java.nio.HeapByteBuffer[pos=0 lim=1023 cap=1024]>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at alluxio.client.file.cache.LocalCacheManagerTest.getOffset(LocalCacheManagerTest.java:490)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:364)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:289)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:237)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:158)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)

beinan commented:
> Error: 9.287 [ERROR] alluxio.client.file.cache.LocalCacheManagerTest.getOffset Time elapsed: 0.123 s <<< FAILURE!
> java.lang.AssertionError: expected:<java.nio.HeapByteBuffer[pos=1 lim=1024 cap=1024]> but was:<java.nio.HeapByteBuffer[pos=0 lim=1023 cap=1024]>
> at org.junit.Assert.fail(Assert.java:89)
> at org.junit.Assert.failNotEquals(Assert.java:835)
> at org.junit.Assert.assertEquals(Assert.java:120)
> at org.junit.Assert.assertEquals(Assert.java:146)
> at alluxio.client.file.cache.LocalCacheManagerTest.getOffset(LocalCacheManagerTest.java:490)
> at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
> at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
> at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
> at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
> at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
> at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
> at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
> at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
> at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
> at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
> at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
> at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
> at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
> at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
> at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:364)
> at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:272)
> at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:237)
> at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:158)
> at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
> at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
> at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
> at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)
> 
> Error: 9.335 [ERROR] alluxio.client.file.cache.LocalCacheManagerTest.getOffset Time elapsed: 0.234 s <<< FAILURE!
> java.lang.AssertionError: expected:<java.nio.HeapByteBuffer[pos=1 lim=1024 cap=1024]> but was:<java.nio.HeapByteBuffer[pos=0 lim=1023 cap=1024]>
> at org.junit.Assert.fail(Assert.java:89)
> at org.junit.Assert.failNotEquals(Assert.java:835)
> at org.junit.Assert.assertEquals(Assert.java:120)
> at org.junit.Assert.assertEquals(Assert.java:146)
> at alluxio.client.file.cache.LocalCacheManagerTest.getOffset(LocalCacheManagerTest.java:490)
> at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
> at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
> at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
> at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
> at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
> at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
> at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
> at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
> at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
> at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
> at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
> at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
> at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
> at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
> at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:364)
> at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:289)
> at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:237)
> at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:158)
> at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:428)
> at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:162)
> at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:562)
> at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:548)

finally, no build failure any more ....  thank you for this error message!   This one is caused by a typo on CacheManager interface, I made put a 0 instead of pass through the offset

apc999 commented:
alluxio-bot, merge this please
